cirq/ops/permutation_gate.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import Counter
+from typing import Any, Dict, Sequence, Tuple, TYPE_CHECKING
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class QubitPermutationGate(raw_types.Gate):
```

Comment:
Out of scope for this PR: I initially thought this would be a gate that permutes computational basis. Instead, it permutes qubits which is a less general operation. I think we should also implement the former, but not necessarily in this PR. I guess the reason to have both is that QubitPermutationGate has an efficient `_apply_unitary_ `.

Reply:
I agree. I was just following almost mechanically @mpharrigan's requirements from #2671 - thus this one is the same as the quirk gate minus the name + id. Having a permutation gate on the computational basis states might be 1) valuable for use cases (although I'm not aware of those just yet) 2) an even better option to consolidate the permutation gate hierarchy in the contrib.aqcuaintance package.

Reply:
Regarding use-cases of the more general permutation gate: unitary corresponding to any bijective function from n-bitstrings to n-bitstrings (including e.g. arithmetic operations) is a permutation matrix. This may not be sufficient grounds for a generic permutation gate since it's reasonable to implement specific gates for such operations instead.

We can defer until we see more use-cases.